pkg/client/unit.go

```diff
@@ -0,0 +1,218 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package client
+
+import (
+	"encoding/json"
+	"sync"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/proto"
+)
+
+// UnitType is the type of the unit, either input or output
+type UnitType proto.UnitType
+
+const (
+	// UnitTypeInput is an input unit.
+	UnitTypeInput = UnitType(proto.UnitType_INPUT)
+	// UnitTypeOutput is an output unit.
+	UnitTypeOutput = UnitType(proto.UnitType_OUTPUT)
+)
+
+// UnitState is the state for the unit, used both for expected and observed state.
+type UnitState proto.State
+
+const (
+	// UnitStateStarting is when a unit is starting.
+	UnitStateStarting = UnitState(proto.State_STARTING)
+	// UnitStateConfiguring is when a unit is currently configuring.
+	UnitStateConfiguring = UnitState(proto.State_CONFIGURING)
+	// UnitStateHealthy is when the unit is working exactly as it should.
+	UnitStateHealthy = UnitState(proto.State_HEALTHY)
+	// UnitStateDegraded is when the unit is working but not exactly as its expected.
+	UnitStateDegraded = UnitState(proto.State_DEGRADED)
+	// UnitStateFailed is when the unit is completely broken and failing to work.
+	UnitStateFailed = UnitState(proto.State_FAILED)
+	// UnitStateStopping is when the unit is stopping.
+	UnitStateStopping = UnitState(proto.State_STOPPING)
+	// UnitStateStopped is when the unit is stopped.
+	UnitStateStopped = UnitState(proto.State_STOPPED)
+)
+
+// Unit represents a distinct item that needs to be operating with-in this process.
+//
+// This is normally N number of inputs and 1 output (possible for multiple in the future).
+type Unit struct {
+	id       string
+	unitType UnitType
+
+	expLock   sync.RWMutex
+	exp       UnitState
+	config    string
```

Comment:
This would be the string representation of the yaml, this seems asking for having his own type? We we can validate the presence of some fields depending on the unit or that we have valid yaml.

Reply:
Sadly we can have a `yaml.RawMessage` because of the way indent works and could break the configuration.

Reply:
Is it YAML, JSON, or something else. The point of the client is that it doesn't know and doesn't care. So its best to keep it just as a string representation that could then be decoded.